dspy/teleprompt/bootstrap_trace.py

```diff
@@ -0,0 +1,298 @@
+        setattr(self._program, name, value)
+
+    def __delattr__(self, name):
+        if name in {"_program", "_call_wrapper"} or "_program" not in self.__dict__:
+            object.__delattr__(self, name)
+            return
+
+        if name in getattr(self, "__dict__", {}):
+            object.__delattr__(self, name)
+        else:
+            delattr(self._program, name)
+
+    # Introspection and representation
+    def __dir__(self):
+        own = set(dir(type(self))) | set(getattr(self, "__dict__", {}).keys())
+        try:
+            prog = set(dir(self._program))
+        except Exception:
+            prog = set()
+        return sorted(own | prog)
+
+    def __repr__(self):
+        return repr(self._program)
+
+    def __str__(self):
+        return str(self._program)
+
+    # Iteration support if the wrapped program is iterable
+    def __iter__(self):
+        return iter(self._program)
+
+    # Equality and hashing delegate to the wrapped object
+    def __eq__(self, other):
+        return self._program == (other._program if isinstance(other, ProgramWrapper) else other)
+
+    def __hash__(self):
+        return hash(self._program)
+
+    # Pickle support
+    def __getstate__(self):
+        # Let Module's __getstate__ do its normal pruning on the wrapper's dict,
+        # but also include the wrapped program and call_wrapper explicitly.
+        state = super().__getstate__()
+        state["_program"] = self._program
+        state["_call_wrapper"] = self._call_wrapper
+        return state
+
+    def __setstate__(self, state):
+        # Restore wrapper internals first.
+        object.__setattr__(self, "_program", state.pop("_program"))
+        object.__setattr__(self, "_call_wrapper", state.pop("_call_wrapper"))
+        # Restore the rest via Module's machinery.
+        super().__setstate__(state)
+
+
+def bootstrap_trace_data(
+    program: Module,
+    dataset: list[Example],
+    metric: Callable | None = None,
+    num_threads: int | None = None,
+    raise_on_error=True,
+    capture_failed_parses=False,
+    failure_score: float = 0,
+    format_failure_score: float = -1,
+    log_format_failures: bool = False,
+) -> list[TraceData]:
+    # Return a list of dicts with the following keys: example_ind, example, prediction, trace, and score
+    # (if metric != None)
+    evaluator = Evaluate(
+        devset=dataset,
+        num_threads=num_threads,
+        display_progress=True,
+        provide_traceback=False,  # TODO(check with team)
+        max_errors=len(dataset) * 10,  # TODO(check with team)
+        failure_score=failure_score,
+    )
+
+    def wrapped_metric(example, prediction, trace=None):
+        prediction, _ = prediction
+        if isinstance(prediction, FailedPrediction):
+            return prediction.format_reward or format_failure_score
+        return metric(example, prediction, trace) if metric else True
+
+    def wrapped_program_callable(program_to_use: Module, **kwargs):
```

Comment:
It might be cleaner to patch this into `program_to_use.forward`, let me check it out and loop back.